ls1intum/Artemis
# `Assessment`: Avoid sending results before the assessment due date has passed

### Checklist
#### General
#### Server

### Motivation and Context
Before the assessment due date, the student should under no circumstances see the manual tutor feedback. However, not every endpoint is currently checking and enforcing this rule.

This refactoring is the first step to consistently check and remove scores and feedback on the server side if the student is not yet allowed to see it.

This PR adds the after assessment due date checks to the `ExerciseDateService` and uses these methods instead of always directly comparing the date with `now()`. This also makes the code more readable, since you now have a descriptive method name in these places and don't need to think yourself about the condition and what is meant by it.

Additionally, it aims to remove the result information from all endpoints if the student should not currently see it.

### Steps for Testing
This PR requires carefull testing and code review!

Please go through the code and check that the logic is still correct (and was not unintentionally inverted).

1. Create exercises (every type) and participate in them as a student.
2. Grade the submissions as a tutor.
3. Check different places on Artemis (e.g. exercise detail, exercise overview, statistics, front page, etc.). Check that the student is never able to see the result generated by the tutor (before the assessment due date).
4. After the assessment due date, check the the result is now shown.

#### Exam Mode Testing
same as above, but with exam exercises. No results should be visible before the "publish exam results" date is reached.

### Review Progress

#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage
newly added lines covered.
